### ISSUE TITLE
LOCK_REQUEST response not expected or in wrong format

### ISSUE THREAD

**Issue by MarcusWichelmann** [imker25/samba_exporter#118]
I'm running Ubuntu 23.04 (Lunar) with the latest Samba version from the Ubuntu repositories: `2:4.17.7+dfsg-1ubuntu2`

When starting `samba_exporter`, it instantly crashes with the following stack trace:
```
# samba_exporter
Error: The response "LOCK_REQUEST: Response for request 3" was not expected or in wrong format
panic: Error: The response "LOCK_REQUEST: Response for request 3" was not expected or in wrong format

goroutine 9 [running]:
tobi.backfrak.de/internal/smbexporterbl/smbexporter.(*SambaExporter).Describe(0xc00007ea50, 0x0?)
	tobi.backfrak.de/internal/smbexporterbl/smbexporter/exporter.go:61 +0x136
github.com/prometheus/client_golang/prometheus.(*Registry).Register.func1()
	github.com/prometheus/client_golang/prometheus/registry.go:280 +0x2b
created by github.com/prometheus/client_golang/prometheus.(*Registry).Register
	github.com/prometheus/client_golang/prometheus/registry.go:279 +0x19b
```

`smbstatus` works just fine:
```
# smbstatus
Samba version 4.17.7-Ubuntu
```

Do you have an idea what could go wrong there? Is this Samba version not supported yet?

Thanks for building this exporter!

**Comment by imker25** [comment]
Hello @MarcusWichelmann 

for now I don't have any idea what is going wrong. But I think we can figure this out if you can provide more verbose log information. Therefor please start both parts with `--verbose`. To do so, add the flag to the `ARGS=''` line in the following two files:
- `/etc/default/samba_exporter`
- `/etc/default/samba_statusd`

After restarting the `samba_exporter` service you should find a lot more log output in your journal. Please send the logs of both services so we can investigate the root cause. 

E. g. run `journalctl -u samba_exporter` and `journalctl -u samba_statusd` to get just the relevant logs.

Greetings
Imker

**Comment by MarcusWichelmann** [comment]
Hi, thanks for your response. I'm sorry, that it took me a bit longer to get back to this.

Here are the relevant logs you requested:

```
-- Boot f7fb00e68c08453bb76e24a8e4936f32 --
systemd[1]: Starting samba_statusd.service - samba-statusd service for prometheus samba-exporter, based on https://github.com/imker25/samba_exporter...
start_samba_statusd[225]: Call:  /usr/bin/samba_statusd -verbose
start_samba_statusd[225]: Version: 1.28.5~ppa1~ubuntu23.04
start_samba_statusd[225]: Verbose: Named pipe for requests: /run/samba_exporter.request.pipe
start_samba_statusd[225]: Verbose: Named pipe for response: /run/samba_exporter.response.pipe
start_samba_statusd[225]: Verbose: Use /usr/bin/smbstatus to get samba status.
start_samba_statusd[225]: Information: Started /usr/bin/samba_statusd, waiting for requests in pipe
start_samba_statusd[225]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
systemd[1]: Started samba_statusd.service - samba-statusd service for prometheus samba-exporter, based on https://github.com/imker25/samba_exporter.
systemd[1]: Starting samba_exporter.service - Prometheus samba-exporter service, based on https://github.com/imker25/samba_exporter...
systemd[1]: Started samba_exporter.service - Prometheus samba-exporter service, based on https://github.com/imker25/samba_exporter.
samba_exporter[286]: Call:  /usr/bin/samba_exporter -web.listen-address=:9922 -verbose
samba_exporter[286]: Version: 1.28.5~ppa1~ubuntu23.04
samba_exporter[286]: Verbose: Named pipe for requests: /run/samba_exporter.request.pipe
samba_exporter[286]: Verbose: Named pipe for response: /run/samba_exporter.response.pipe
samba_exporter[286]: Verbose: Setup prometheus exporter
samba_exporter[286]: Verbose: Request samba_statusd to get prometheus descriptions
samba_exporter[286]: Verbose: Send "PROCESS_REQUEST:" request with ID 1 on pipe
start_samba_statusd[225]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[225]: Verbose: Handle "PROCESS_REQUEST:" with id 1
samba_exporter[286]: Verbose: Wait for "PROCESS_REQUEST:" response with ID 1 on pipe
samba_exporter[286]: Verbose: Handle "PROCESS_REQUEST:" response with ID 1 from pipe
samba_exporter[286]: Verbose: Send "SHARE_REQUEST:" request with ID 2 on pipe
start_samba_statusd[225]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[225]: Verbose: Handle "SHARE_REQUEST:" with id 2
samba_exporter[286]: Verbose: Wait for "SHARE_REQUEST:" response with ID 2 on pipe
samba_exporter[286]: Verbose: Handle "SHARE_REQUEST:" response with ID 2 from pipe
samba_exporter[286]: Verbose: Send "LOCK_REQUEST:" request with ID 3 on pipe
samba_exporter[286]: Verbose: Wait for "LOCK_REQUEST:" response with ID 3 on pipe
start_samba_statusd[225]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[225]: Verbose: Handle "LOCK_REQUEST:" with id 3
samba_exporter[286]: Verbose: Handle "LOCK_REQUEST:" response with ID 3 from pipe
samba_exporter[286]: Error: The response "LOCK_REQUEST: Response for request 3" was not expected or in wrong format
samba_exporter[286]: panic: Error: The response "LOCK_REQUEST: Response for request 3" was not expected or in wrong format
samba_exporter[286]: goroutine 20 [running]:
samba_exporter[286]: tobi.backfrak.de/internal/smbexporterbl/smbexporter.(*SambaExporter).Describe(0xc00013a000, 0x0?)
samba_exporter[286]:         tobi.backfrak.de/internal/smbexporterbl/smbexporter/exporter.go:61 +0x136
samba_exporter[286]: github.com/prometheus/client_golang/prometheus.(*Registry).Register.func1()
samba_exporter[286]:         github.com/prometheus/client_golang/prometheus/registry.go:280 +0x2b
samba_exporter[286]: created by github.com/prometheus/client_golang/prometheus.(*Registry).Register
samba_exporter[286]:         github.com/prometheus/client_golang/prometheus/registry.go:279 +0x19b
start_samba_statusd[225]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
systemd[1]: samba_exporter.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
systemd[1]: samba_exporter.service: Failed with result 'exit-code'.

```

`smbd`, `samba_statusd` and `samba_exporter` are all running in the same LXC container. It ran fine, until I updated the LXC from Ubuntu 22.04 to 23.04 and, along that, samba `4.16` to `4.17.7`.

I noticed, that, when I grab a coffee and wait a few minutes, and then try to start `samba_exporter` again, it runs just fine. This seems to be releated to clients reconnecting to the server and causing entries in the `Locked files` list of `smbstatus`.  As soon as the `Locked files` list isn't empty anymore, `samba_exporter` doesn't crash. When it's empty again, `samba_exporter` crashes again. This is reproducible.

So, probably something changed regarding the representation of an empty `LOCK_REQUEST` response?

**Comment by imker25** [comment]
@MarcusWichelmann 

thank you for all that information. So I think you are right, if it only happens when the `LOCK_REQUEST` response is empty, then there must be a change in the output of `smbstatus -L -n`.

I'm pretty sure I can adapt the code so it will work fine if you can attache the output of this call to this bug. So please run the following when no clients are connected and no looks are open, and then add ` /tmp/smbstatus-lock-out.txt ` to this issue:

```bash
smbstatus -L -n > /tmp/smbstatus-lock-out.txt 2>&1
```

I will add this file to my set of test data, and fix the code as needed
Thank you
imker

**Comment by imker25** [comment]
Yes, it looks like the new version adds a newline at the end of the message, what the old one didn't.

I checked in a fix for this issue with 8fe101bad45fdff5c7a45ba19e3c4cb208fad3fe .
Can you please install the [V1.29.2-pre](https://github.com/imker25/samba_exporter/releases/tag/1.29.2-pre) and test again. This version should solve your problems.

**Comment by MarcusWichelmann** [comment]
Thanks for the quick fix! Unfortunately, the issue doesn't seem to be resolved:

```
# samba_statusd -print-version
Version: 1.29.2-pre~ppa1~ubuntu23.04
# samba_exporter -print-version
Version: 1.29.2-pre~ppa1~ubuntu23.04
# smbstatus -L -n
No locked files

# samba_exporter -verbose
Call:  samba_exporter -verbose
Version: 1.29.2-pre~ppa1~ubuntu23.04
Verbose: Named pipe for requests: /run/samba_exporter.request.pipe
Verbose: Named pipe for response: /run/samba_exporter.response.pipe
Verbose: Setup prometheus exporter
Verbose: Request samba_statusd to get prometheus descriptions
Verbose: Send "PROCESS_REQUEST:" request with ID 1 on pipe
Verbose: Wait for "PROCESS_REQUEST:" response with ID 1 on pipe
Verbose: Handle "PROCESS_REQUEST:" response with ID 1 from pipe
Verbose: Send "SHARE_REQUEST:" request with ID 2 on pipe
Verbose: Wait for "SHARE_REQUEST:" response with ID 2 on pipe
Verbose: Handle "SHARE_REQUEST:" response with ID 2 from pipe
Verbose: Send "LOCK_REQUEST:" request with ID 3 on pipe
Verbose: Wait for "LOCK_REQUEST:" response with ID 3 on pipe
Verbose: Handle "LOCK_REQUEST:" response with ID 3 from pipe
Error: The response "LOCK_REQUEST: Response for request 3" was not expected or in wrong format
panic: Error: The response "LOCK_REQUEST: Response for request 3" was not expected or in wrong format

goroutine 10 [running]:
tobi.backfrak.de/internal/smbexporterbl/smbexporter.(*SambaExporter).Describe(0xc00007eaa0, 0x0?)
	tobi.backfrak.de/internal/smbexporterbl/smbexporter/exporter.go:61 +0x136
github.com/prometheus/client_golang/prometheus.(*Registry).Register.func1()
	github.com/prometheus/client_golang/prometheus/registry.go:280 +0x2b
created by github.com/prometheus/client_golang/prometheus.(*Registry).Register
	github.com/prometheus/client_golang/prometheus/registry.go:279 +0x19b

```

I installed the `samba-exporter_1.29.2-pre.ppa1.ubuntu23.04_amd64.deb`.

Very strange. ;)

**Comment by MarcusWichelmann** [comment]
Here is the full log again:

```
systemd[1]: Starting samba_statusd.service - samba-statusd service for prometheus samba-exporter, based on https://github.com/imker25/samba_exporter...
start_samba_statusd[266]: Call:  /usr/bin/samba_statusd -verbose
start_samba_statusd[266]: Version: 1.29.2-pre~ppa1~ubuntu23.04
start_samba_statusd[266]: Verbose: Named pipe for requests: /run/samba_exporter.request.pipe
start_samba_statusd[266]: Verbose: Named pipe for response: /run/samba_exporter.response.pipe
start_samba_statusd[266]: Verbose: Use /usr/bin/smbstatus to get samba status.
start_samba_statusd[266]: Information: Started /usr/bin/samba_statusd, waiting for requests in pipe
start_samba_statusd[266]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
systemd[1]: Started samba_statusd.service - samba-statusd service for prometheus samba-exporter, based on https://github.com/imker25/samba_exporter.
systemd[1]: Starting samba_exporter.service - Prometheus samba-exporter service, based on https://github.com/imker25/samba_exporter...
systemd[1]: Started samba_exporter.service - Prometheus samba-exporter service, based on https://github.com/imker25/samba_exporter.
samba_exporter[324]: Call:  /usr/bin/samba_exporter -web.listen-address=:9922 -verbose
samba_exporter[324]: Version: 1.29.2-pre~ppa1~ubuntu23.04
samba_exporter[324]: Verbose: Named pipe for requests: /run/samba_exporter.request.pipe
samba_exporter[324]: Verbose: Named pipe for response: /run/samba_exporter.response.pipe
samba_exporter[324]: Verbose: Setup prometheus exporter
samba_exporter[324]: Verbose: Request samba_statusd to get prometheus descriptions
samba_exporter[324]: Verbose: Send "PROCESS_REQUEST:" request with ID 1 on pipe
samba_exporter[324]: Verbose: Wait for "PROCESS_REQUEST:" response with ID 1 on pipe
start_samba_statusd[266]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[266]: Verbose: Handle "PROCESS_REQUEST:" with id 1
samba_exporter[324]: Verbose: Handle "PROCESS_REQUEST:" response with ID 1 from pipe
samba_exporter[324]: Verbose: Send "SHARE_REQUEST:" request with ID 2 on pipe
samba_exporter[324]: Verbose: Wait for "SHARE_REQUEST:" response with ID 2 on pipe
start_samba_statusd[266]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[266]: Verbose: Handle "SHARE_REQUEST:" with id 2
samba_exporter[324]: Verbose: Handle "SHARE_REQUEST:" response with ID 2 from pipe
samba_exporter[324]: Verbose: Send "LOCK_REQUEST:" request with ID 3 on pipe
samba_exporter[324]: Verbose: Wait for "LOCK_REQUEST:" response with ID 3 on pipe
start_samba_statusd[266]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[266]: Verbose: Handle "LOCK_REQUEST:" with id 3
samba_exporter[324]: Verbose: Handle "LOCK_REQUEST:" response with ID 3 from pipe
samba_exporter[324]: Error: The response "LOCK_REQUEST: Response for request 3" was not expected or in wrong format
samba_exporter[324]: panic: Error: The response "LOCK_REQUEST: Response for request 3" was not expected or in wrong format
samba_exporter[324]: goroutine 7 [running]:
samba_exporter[324]: tobi.backfrak.de/internal/smbexporterbl/smbexporter.(*SambaExporter).Describe(0xc00007c050, 0x0?)
samba_exporter[324]:         tobi.backfrak.de/internal/smbexporterbl/smbexporter/exporter.go:61 +0x136
samba_exporter[324]: github.com/prometheus/client_golang/prometheus.(*Registry).Register.func1()
samba_exporter[324]:         github.com/prometheus/client_golang/prometheus/registry.go:280 +0x2b
samba_exporter[324]: created by github.com/prometheus/client_golang/prometheus.(*Registry).Register
samba_exporter[324]:         github.com/prometheus/client_golang/prometheus/registry.go:279 +0x19b

```

**Comment by imker25** [comment]
@MarcusWichelmann 

I think I now found out why the first attempt to fix the issue did not work. The line `Error: The response "LOCK_REQUEST: Response for request 3" was not expected or in wrong format` tells us, that the answer of `samba_statusd` to our request was just `LOCK_REQUEST: Response for request 3`. But the expected answer would be something like shown below:

```txt
LOCK_REQUEST: Response for request 3
No locked files

```

So for some reason, the output of `smbstatus -L -n` is an empty string whenever you run into that error case.

I'm not sure, how I want to handle this, since for me it seems like a bug in `smbstatus`.   `smbstatus` should return always the same message for the same call, not one time an empty string and the other time `No locked files`. Can it be, that the output of  `smbstatus` changes when `smbd` is running or not? 

Could you run `smbstatus -L -n` before the samba daemon was started on your system and see if you get no output from that call in this case?

**Comment by imker25** [comment]
@MarcusWichelmann 

I now have reworked the handling of empty response data, as it is shown in the logs provided by you. 
[V1.29.6-pre](https://github.com/imker25/samba_exporter/releases/tag/1.29.6-pre) should be able to deal with it without crashing. Can you please test this version again?

**Comment by MarcusWichelmann** [comment]
@imker25 

Just tried it and it seems to work now. I cannot reproduce the issue any more. 🎉 

Letting the exporter running for a while now and unsilenced the monitoring for it to see if everything works fine.

Thank you a lot for the fix!

**Comment by MarcusWichelmann** [comment]
BTW, here is the new log output:

```
systemd[1]: Starting samba_statusd.service - samba-statusd service for prometheus samba-exporter, based on https://github.com/imker25/samba_exporter...
start_samba_statusd[223]: Call:  /usr/bin/samba_statusd -verbose
start_samba_statusd[223]: Version: 1.29.6-pre~ppa1~ubuntu23.04
start_samba_statusd[223]: Verbose: Named pipe for requests: /run/samba_exporter.request.pipe
start_samba_statusd[223]: Verbose: Named pipe for response: /run/samba_exporter.response.pipe
start_samba_statusd[223]: Verbose: Use /usr/bin/smbstatus to get samba status.
start_samba_statusd[223]: Information: Started /usr/bin/samba_statusd, waiting for requests in pipe
start_samba_statusd[223]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
systemd[1]: Started samba_statusd.service - samba-statusd service for prometheus samba-exporter, based on https://github.com/imker25/samba_exporter.
systemd[1]: Starting samba_exporter.service - Prometheus samba-exporter service, based on https://github.com/imker25/samba_exporter...
systemd[1]: Started samba_exporter.service - Prometheus samba-exporter service, based on https://github.com/imker25/samba_exporter.
samba_exporter[281]: Call:  /usr/bin/samba_exporter -web.listen-address=:9922 -verbose
samba_exporter[281]: Version: 1.29.6-pre~ppa1~ubuntu23.04
samba_exporter[281]: Verbose: Named pipe for requests: /run/samba_exporter.request.pipe
samba_exporter[281]: Verbose: Named pipe for response: /run/samba_exporter.response.pipe
samba_exporter[281]: Verbose: Setup prometheus exporter
samba_exporter[281]: Verbose: Request samba_statusd to get prometheus descriptions
samba_exporter[281]: Verbose: Send "PROCESS_REQUEST:" request with ID 1 on pipe
samba_exporter[281]: Verbose: Wait for "PROCESS_REQUEST:" response with ID 1 on pipe
start_samba_statusd[223]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[223]: Verbose: Handle "PROCESS_REQUEST:" with id 1
samba_exporter[281]: Verbose: Handle "PROCESS_REQUEST:" response with ID 1 from pipe
samba_exporter[281]: Verbose: Send "SHARE_REQUEST:" request with ID 2 on pipe
samba_exporter[281]: Verbose: Wait for "SHARE_REQUEST:" response with ID 2 on pipe
start_samba_statusd[223]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[223]: Verbose: Handle "SHARE_REQUEST:" with id 2
samba_exporter[281]: Verbose: Handle "SHARE_REQUEST:" response with ID 2 from pipe
samba_exporter[281]: Verbose: Send "LOCK_REQUEST:" request with ID 3 on pipe
samba_exporter[281]: Verbose: Wait for "LOCK_REQUEST:" response with ID 3 on pipe
start_samba_statusd[223]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[223]: Verbose: Handle "LOCK_REQUEST:" with id 3
samba_exporter[281]: Verbose: Handle "LOCK_REQUEST:" response with ID 3 from pipe
samba_exporter[281]: Verbose: Send "PS_REQUEST:" request with ID 4 on pipe
samba_exporter[281]: Verbose: Wait for "PS_REQUEST:" response with ID 4 on pipe
samba_exporter[281]: Information: Got an empty string from 'smbstatus -L -n'
start_samba_statusd[223]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[223]: Verbose: Handle "PS_REQUEST:" with id 4
samba_exporter[281]: Verbose: Handle "PS_REQUEST:" response with ID 4 from pipe
samba_exporter[281]: Verbose: Got an empty share table when requesting "smbstatus -S -n" from samba_statusd
samba_exporter[281]: Verbose: Got an empty process table when requesting "smbstatus -p -n" from samba_statusd
samba_exporter[281]: Verbose: Handle samba_statusd response and set prometheus descriptions
samba_exporter[281]: Information: Started /usr/bin/samba_exporter, get metrics on http://:9922/metrics
samba_exporter[281]: Verbose: Request samba_statusd to get prometheus metrics
samba_exporter[281]: Verbose: Send "PROCESS_REQUEST:" request with ID 5 on pipe
samba_exporter[281]: Verbose: Wait for "PROCESS_REQUEST:" response with ID 5 on pipe
start_samba_statusd[223]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[223]: Verbose: Handle "PROCESS_REQUEST:" with id 5
samba_exporter[281]: Verbose: Handle "PROCESS_REQUEST:" response with ID 5 from pipe
samba_exporter[281]: Verbose: Send "SHARE_REQUEST:" request with ID 6 on pipe
samba_exporter[281]: Verbose: Wait for "SHARE_REQUEST:" response with ID 6 on pipe
start_samba_statusd[223]: Verbose: Wait for requests in: /run/samba_exporter.request.pipe
start_samba_statusd[223]: Verbose: Handle "SHARE_REQUEST:" with id 6
samba_exporter[281]: Verbose: Handle "SHARE_REQUEST:" response with ID 6 from pipe
samba_exporter[281]: Verbose: Send "LOCK_REQUEST:" request with ID 7 on pipe
samba_exporter[281]: Verbose: Wait for "LOCK_REQUEST:" response with ID 7 on pipe

...
```

**Comment by imker25** [comment]
@MarcusWichelmann 

Thank you for the confirmation that [V1.29.6-pre](https://github.com/imker25/samba_exporter/releases/tag/1.29.6-pre) fixes the issue. I will release a the 1.29.X version (not -pre) within the next days and then close this bug.

**Comment by imker25** [comment]
Fixed with [V1.29.6](https://github.com/imker25/samba_exporter/releases/tag/1.29.6)
